### DISCUSSION TITLE
How to schedule a Task to be executed under the Users group from a Windows C# Service?

### DISCUSSION THREAD

**Discussion by scornz** [dahall/TaskScheduler#893] [Q&A] [answered]
I have a WPF application that I am attempting to start upon every logon of a user. The only issue is that the specified group is "SYSTEM" since I am creating the task from a service application. So, when the application is started, the UI will never appear (as expected). If I go into the Task Scheduler UI and manually change the user/group to "Users", the WPF application opens and the UI is visible. This is the current script:
```C#
// Create a new task definition for the local machine and assign properties
TaskDefinition td = TaskService.Instance.NewTask();
td.RegistrationInfo.Description = "Launch test-login";
//td.Settings.RunOnlyIfLoggedOn = true;
#pragma warning disable CA1416 // Validate platform compatibility
string userId = System.Security.Principal.WindowsIdentity.GetCurrent().Name;
#pragma warning restore CA1416 // Validate platform compatibility

td.Principal.UserId = userId;
td.Principal.LogonType = TaskLogonType.InteractiveToken;

LogonTrigger lt = new LogonTrigger();
td.Triggers.Add(lt);

string executePath = Path.GetDirectoryName(Environment.GetCommandLineArgs()[0]) + "\\login.exe";
td.Actions.Add(executePath);

// Register the task in the root folder of the local machine
TaskService.Instance.RootFolder.RegisterTaskDefinition("login-start", td, TaskCreation.CreateOrUpdate, userId, logonType: TaskLogonType.InteractiveToken);
```

I understand why the above code _won't_ work, since `GetCurrent()` will simply return "SYSTEM" since this code is in a service. Any advice on how to solve this?

**Comment by dahall** [comment, marked as answer]
You're on the right track. Do not set the `td.Principal` properties. Then, in the `RegisterTaskDefinition` call, do the following:
```cs
TaskService.Instance.RootFolder.RegisterTaskDefinition("westlightlogin-start", td, TaskCreation.CreateOrUpdate, "YourDomain\\GroupAcct", null, TaskLogonType.Group);
```

**Reply by scornz** [reply]
Ah, okay. That makes sense. In order to make this universally supported, how would I get the domain name of the "Users" group? This may be a stupid question, but it just seems quite difficult from a windows service.

**Reply by dahall** [reply]
Try "BUILTIN\Authenticated Users". I think that's the format to get the special group for all users.

**Comment by scornz** [comment]
Removing the `td.Principal` properties and modifying the `RegisterTaskDefinition` call works wonders. The group is set to "Authenticated Users" and the WPF application UI is now visible. Here is what the call ended up being:

```C#
// Register the task in the root folder of the local machine
TaskService.Instance.RootFolder.RegisterTaskDefinition("login-start", td, TaskCreation.CreateOrUpdate, "Authenticated Users", null, TaskLogonType.Group);
```
